src/llama-kv-cache-hybrid-recurrent.h

```diff
@@ -0,0 +1,144 @@
+#pragma once
+
+#include "llama-batch.h"
+#include "llama-graph.h"
+#include "llama-kv-cache-recurrent.h"
+#include "llama-kv-cache-unified.h"
+#include "llama-kv-cells.h"
+#include "llama-memory.h"
+
+#include <memory>
+#include <vector>
+
+//
+// llama_kv_cache_hybrid_recurrent
+//
+
+// utilizes instances of llama_kv_cache_recurrent and llama_kv_cache_unified to
+//   support models where each layer may be either attention-based or recurrent
+
+class llama_kv_cache_hybrid_recurrent : public llama_memory_i {
+public:
+    llama_kv_cache_hybrid_recurrent(
+            const llama_model & model,
+                                /* attn */
+                    ggml_type   attn_type_k,
+                    ggml_type   attn_type_v,
+                         bool   attn_v_trans,
+                     uint32_t   attn_kv_size,
+                     uint32_t   attn_n_pad,
+                     uint32_t   attn_n_swa,
+               llama_swa_type   attn_swa_type,
+                                /* recurrent */
+                    ggml_type   recurrent_type_k,
+                    ggml_type   recurrent_type_v,
+                     uint32_t   recurrent_kv_size,
+                                /* common */
+                     uint32_t   n_seq_max,
+                         bool   offload);
+
+    ~llama_kv_cache_hybrid_recurrent() = default;
+
+    //
+    // llama_memory_i
+    //
+
+    llama_memory_state_ptr init_batch(
+            const llama_batch & batch,
+            uint32_t n_ubatch,
+            bool embd_pooled,
+            bool logits_all) override;
+
+    llama_memory_state_ptr init_full() override;
+
+    llama_memory_state_ptr init_update(llama_context * lctx, bool optimize) override;
+
+    bool get_can_shift() const override;
+
+    void clear(bool data) override;
+
+    bool seq_rm  (llama_seq_id seq_id,                              llama_pos p0, llama_pos p1) override;
+    void seq_cp  (llama_seq_id seq_id_src, llama_seq_id seq_id_dst, llama_pos p0, llama_pos p1) override;
+    void seq_keep(llama_seq_id seq_id)                                                          override;
+    void seq_add (llama_seq_id seq_id,                              llama_pos p0, llama_pos p1, llama_pos shift) override;
+    void seq_div (llama_seq_id seq_id,                              llama_pos p0, llama_pos p1, int d) override;
+
+    llama_pos seq_pos_min(llama_seq_id seq_id) const override;
+    llama_pos seq_pos_max(llama_seq_id seq_id) const override;
+
+    // state write/load
+
+    void state_write(llama_io_write_i & io, llama_seq_id seq_id = -1) const override;
+    void state_read (llama_io_read_i  & io, llama_seq_id seq_id = -1)       override;
+
+    //
+    // llama_kv_cache_hybrid_recurrent specific API
+    //
+
+    llama_kv_cache_unified   * get_kv_attn     () const;
```

Comment:
I don't think I've ever actually used the `friend` keyword in real code, but I think we _could_ make `llama_kv_cache_hybrid_recurrent_state` a `friend` and then make these `private` for slightly better encapsulation.

Reply:
Actually, if we made it a `friend`, these methods could just be removed since the child `unique_ptr` instances could just be accessed directly.